File: backend/app/services/audit_log_service.py
"""Audit log service. Records every important business action."""
from __future__ import annotations

import logging
from typing import List, Optional

from sqlalchemy.orm import Session

from app.models.audit_log import AuditAction, AuditLog
from app.repositories.audit_log_repository import AuditLogRepository


logger = logging.getLogger(__name__)


class AuditLogService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.repo = AuditLogRepository(db)

    def log_action(
        self,
        *,
        user_id: Optional[int],
        action: AuditAction,
        entity_type: str,
        entity_id: Optional[int] = None,
        details: Optional[str] = None,
    ) -> AuditLog:
        log = AuditLog(
            user_id=user_id,
            action=action,
            entity_type=entity_type,
            entity_id=entity_id,
            details=details,
        )
        self.repo.add(log)
        logger.info(
            "audit user_id=%s action=%s entity_type=%s entity_id=%s",
            user_id,
            action.value,
            entity_type,
            entity_id,
        )
        return log

    def list_logs(
        self,
        action: Optional[AuditAction] = None,
        limit: int = 100,
    ) -> List[AuditLog]:
        return self.repo.list_logs(action=action, limit=limit)

File: backend/src/main/java/com/example/bwms/service/AuditLogService.java
// Translated from: backend/app/services/audit_log_service.py
package com.example.bwms.service;

import com.example.bwms.model.AuditAction;
import com.example.bwms.model.AuditLog;
import com.example.bwms.repository.AuditLogRepository;
import jakarta.persistence.criteria.JoinType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AuditLogService {

    private static final Logger log = LoggerFactory.getLogger(AuditLogService.class);

    private final AuditLogRepository auditLogRepository;

    public AuditLogService(AuditLogRepository auditLogRepository) {
        this.auditLogRepository = auditLogRepository;
    }

    // Python: log_action(*, user_id, action, entity_type, entity_id, details)
    public AuditLog logAction(@Nullable Long userId, AuditAction action,
                              String entityType, @Nullable Long entityId,
                              @Nullable String details) {
        AuditLog entry = new AuditLog(userId, action, entityType, entityId, details);
        auditLogRepository.save(entry);
        log.info("audit user_id={} action={} entity_type={} entity_id={}",
                userId, action.getValue(), entityType, entityId);
        return entry;
    }

    public List<AuditLog> listLogs(@Nullable AuditAction action, int limit) {
        Specification<AuditLog> spec = (root, query, cb) -> {
            if (query.getResultType() == AuditLog.class) {
                root.fetch("user", JoinType.LEFT);
            }
            return action == null ? cb.conjunction() : cb.equal(root.get("action"), action);
        };
        return auditLogRepository.findAll(
                spec,
                PageRequest.of(0, limit, Sort.by(Sort.Direction.DESC, "createdAt"))
        ).getContent();
    }
}
